fission-web-server/library/Fission/Web/Server/IPFS/Cluster.hs

```diff
@@ -1,7 +1,45 @@
 module Fission.Web.Server.IPFS.Cluster
-  ( module Fission.Web.Server.IPFS.Cluster.Client
-  , module Fission.Web.Server.IPFS.Cluster.Error
+  ( pin
+  , unpin
+  , pinStream
+  , module Fission.Web.Server.IPFS.Cluster.Class
   ) where
 
-import           Fission.Web.Server.IPFS.Cluster.Client
-import           Fission.Web.Server.IPFS.Cluster.Error
+-- 🌐
+
+import           Network.IPFS.CID.Types
+import qualified Network.IPFS.Client                         as IPFS
+
+import           Servant.Client
+import qualified Servant.Client.Streaming                    as Streaming
+
+-- ⚛️
+
+import           Fission.Prelude
+
+import           Fission.Web.Async
+import           Fission.Web.Server.IPFS.Cluster.Class
+
+import           Fission.Web.Server.IPFS.Streaming.Pin.Types
+
+pin :: MonadIPFSCluster m => CID -> m (Either ClientError ())
+pin (CID hash) = do
+  asyncRefs <- runCluster $ IPFS.pin hash
+  waitAnySuccessCatch asyncRefs >>= \case
+    Left err -> return $ Left err
+    Right _  -> return $ Right ()
+
+unpin :: MonadIPFSCluster m => CID -> m (Either ClientError ())
+unpin (CID hash) = do
+  asyncRefs <- runCluster $ IPFS.unpin hash True -- Recursive flag
+  waitAnySuccessCatch asyncRefs >>= \case
+    Left err -> return $ Left err
+    Right _  -> return $ Right ()
+
+pinStream :: MonadIPFSCluster m => CID -> m (Either ClientError ())
+pinStream cid = do
+  pseudoStreams <- streamCluster $ (Streaming.client $ Proxy @PinComplete) (Just cid) (Just True)
+  let asyncRefs = fst <$> pseudoStreams
+  waitAnySuccessCatch asyncRefs >>= \case
+    Left err -> return $ Left err
+    Right _  -> return $ Right ()
```

Comment:
So that we can do the idle timeout (i.e. requires getting streaming updates from the request)

Reply:
:+1: this seems like it also gets us like 80% of the way to streaming progress back to the client

Reply:
Yeah, it does. The only thing that we're really missing for that is it changes the contract, so we need new endpoints to do streaming.

When we do that, we may want to explore proper streaming as well (e.g. [streamly](https://hackage.haskell.org/package/streamly))